becquerel/core/fitting.py

```diff
@@ -0,0 +1,871 @@
+
+    def _make_model(self, model):
+        if isstring(model) or isinstance(model, Model):
+            model = [model]
+        # Convert the model(s) to a list of Model classes / Model instancess
+        self._model_cls_cnt = {}
+        model_translated = [self._translate_model(m) for m in model]
+        # Build complete model with appropriate prefixes
+        model_prefixes = set()
+        models = []
+        name = ''
+        for m in model_translated:
+            if inspect.isclass(m):
+                prefix_base = m.__name__.lower()
+                if prefix_base.endswith('model'):
+                    prefix_base = prefix_base[:-5]
+                if self._model_cls_cnt[m] == 1:
+                    prefix = '{}_'.format(prefix_base)
+                else:
+                    for i in range(self._model_cls_cnt[m]):
+                        prefix = '{}{}_'.format(prefix_base, i)
+                        if prefix not in model_prefixes:
+                            break
+                m_instance = m(prefix=prefix)
+            else:
+                m_instance = m
+            if m_instance.prefix in model_prefixes:
+                raise FittingError(
+                    'A model prefix is not unique: ' +
+                    '{} '.format(m_instance.prefix) +
+                    'All models: {}'.format(model_translated)
+                )
+            model_prefixes.add(m_instance.prefix)
+            models.append(m_instance)
+            name += m_instance._name.capitalize()
+        # Construct final model
+        self._name = name
+        self._model = models[0]
+        for m in models[1:]:
+            self._model += m
+
+    def _guess_param_defaults(self, **kwargs):
+        params = []
+        for comp in self.model.components:
+            params += comp.guess(self.y_roi, x=self.x_roi, dx=self.dx_roi)
+        return params
+
+    def guess_param_defaults(self, update=False, **kwargs):
+        defaults = self._guess_param_defaults(**kwargs)
+        if update:
+            # TODO: check this logic
+            if defaults is not None:
+                for dp in defaults:
+                    self.set_param(*dp)
+        return defaults
+
+    def fit(self, backend='lmfit'):
+        assert self.y is not None, \
+            'No data initialized, did you call set_data?'
+        self.result = None
+        y_roi_norm = self.y_roi
+        if self.dx is not None:
+            y_roi_norm = self.y_roi / self.dx_roi
+        if backend.lower().strip() == 'lmfit':
+            weights = self.y_unc_roi ** -1.0
+            self.result = self.model.fit(
+                y_roi_norm, self.params, x=self.x_roi, weights=weights)
+        elif backend.lower().strip() == 'lmfit-pml':
+            self._set_likelihood_residual()
+            self.result = self.model.fit(
+                self.y_roi, self.params, #self.result.params,
+                x=self.x_roi, weights=self.dx_roi,
```

Comment:
Are the weights correct here? If so, let's add a note why.

Reply:
> @jccurtis those are with the old guess function. We can look into improving guesses in the future.

Will be resolved in #225

Reply:
- [ ] Add a note that this is converting to counts since the inverse of the weights are applied